lib/class-wp-duotone-gutenberg.php

```diff
@@ -0,0 +1,375 @@
+
+			$duotone_attr = $block['attrs']['style']['color']['duotone'];
+			$is_preset    = is_string( $duotone_attr ) && WP_Duotone_Gutenberg::is_preset( $duotone_attr );
+			$is_css       = is_string( $duotone_attr ) && ! $is_preset;
+			$is_custom    = is_array( $duotone_attr );
+
+			if ( $is_preset ) {
+
+				// Extract the slug from the preset variable string.
+				$slug = WP_Duotone_Gutenberg::gutenberg_get_slug_from_attr( $duotone_attr );
+
+				// Utilize existing preset CSS custom property.
+				$declaration_value = WP_Duotone_Gutenberg::get_css_var( $slug );
+
+				WP_Duotone_Gutenberg::$output[ $slug ] = WP_Duotone_Gutenberg::$global_styles_presets[ $slug ];
+
+			} elseif ( $is_css ) {
+				// Build a unique slug for the filter based on the CSS value.
+				$slug = wp_unique_id( sanitize_key( $duotone_attr . '-' ) );
+
+				// Pass through the CSS value.
+				$declaration_value = $duotone_attr;
+			} elseif ( $is_custom ) {
+				// Build a unique slug for the filter based on the array of colors.
+				$slug = wp_unique_id( sanitize_key( implode( '-', $duotone_attr ) . '-' ) );
+
+				$filter_data = array(
+					'slug'   => $slug,
+					'colors' => $duotone_attr,
+				);
+				// Build a customized CSS filter property for unique slug.
+				$declaration_value = gutenberg_get_duotone_filter_property( $filter_data );
+
+				WP_Duotone_Gutenberg::$output[ $slug ] = $filter_data;
+			}
+		} elseif ( $has_global_styles_duotone ) {
+			$slug = WP_Duotone_Gutenberg::$global_styles_block_names[ $block['blockName'] ];
+
+			// Utilize existing preset CSS custom property.
+			$declaration_value = WP_Duotone_Gutenberg::get_css_var( $slug );
+
+			WP_Duotone_Gutenberg::$output[ $slug ] = WP_Duotone_Gutenberg::$global_styles_presets[ $slug ];
+		}
+
+		// - Applied as a class attribute to the block wrapper.
+		// - Used as a selector to apply the filter to the block.
+		$filter_id = gutenberg_get_duotone_filter_id( array( 'slug' => $slug ) );
+
+		// Build the CSS selectors to which the filter will be applied.
+		$selector = WP_Theme_JSON_Gutenberg::scope_selector( '.' . $filter_id, $duotone_support );
+
+		// We only want to add the selector if we have it in the output already, essentially skipping 'unset'.
+		if ( array_key_exists( $slug, WP_Duotone_Gutenberg::$output ) ) {
+			WP_Duotone_Gutenberg::$output[ $slug ]['selector'] = $selector;
+		}
+
+		// Calling gutenberg_style_engine_get_stylesheet_from_css_rules ensures that
+		// the styles are rendered in an inline for block supports because we're
+		// using the `context` option to instruct it so.
+		gutenberg_style_engine_get_stylesheet_from_css_rules(
+			array(
+				array(
+					'selector'     => $selector,
+					'declarations' => array(
+						// !important is needed because these styles
+						// render before global styles,
+						// and they should be overriding the duotone
+						// filters set by global styles.
+						'filter' => $declaration_value . ' !important',
```

Comment:
This comment is now out of date since we aren't generating duotone styles in global styles anymore.

Reply:
It seems like we should be able to, but when I remove the `!important`, the CSS rule isn't output on the page at all. I'm not sure why though. Something is stripping it out or not allowing it be added. Looking into it.

Reply:
That's related to the kses filter being especially specific for this feature in https://github.com/WordPress/gutenberg/pull/48281/files/08d51b9dd8edc80f32bc035f7f8b6e5931508b70..8dddbd21f2aa9a5224e5214ebe3e9cfee82d386b#diff-9cac08004c469d02cbe46472d33dfb2349d2694d223be7bad4a59d4204a1702d until the more general `filter` property is added to the list of kses properties that support `url()` in https://github.com/WordPress/wordpress-develop/pull/4108.